packages/api/src/utils/login.ts

```diff
@@ -228,9 +236,45 @@ export type SubscriptionInfo = {
   tenantId: string;
 };
 
+export declare type TokenRequest = {
+  scopes: Array<string>;
+  showDialog?: boolean;
+};
+export type LoginStatus = {
+  status: string;
+  token?: string;
+  accountInfo?: Record<string, unknown>;
+};
+export interface M365TokenProvider {
+  getAccessToken(tokenRequest: TokenRequest): Promise<Result<string, FxError>>;
+  getJsonObject(tokenRequest: TokenRequest): Promise<Result<Record<string, unknown>, FxError>>;
+  getStatus(tokenRequest: TokenRequest): Promise<Result<LoginStatus, FxError>>;
+  setStatusChangeMap(
+    name: string,
+    tokenRequest: TokenRequest,
+    statusChange: (
+      status: string,
+      token?: string,
+      accountInfo?: Record<string, unknown>
+    ) => Promise<void>,
+    immediateCall?: boolean
+  ): Promise<Result<boolean, FxError>>;
+  removeStatusChangeMap(name: string): Promise<Result<boolean, FxError>>;
+}
+
 export type TokenProvider = {
   azureAccountProvider: AzureAccountProvider;
+  /**
+   * @deprecated The method should not be used, please update to M365TokenProvider
+   */
   graphTokenProvider: GraphTokenProvider;
+  /**
+   * @deprecated The method should not be used, please update to M365TokenProvider
+   */
   appStudioToken: AppStudioTokenProvider;
+  /**
+   * @deprecated The method should not be used, please update to M365TokenProvider
+   */
   sharepointTokenProvider: SharepointTokenProvider;
+  m365TokenProvider: M365TokenProvider;
```

Comment:
For further upgrading, any side-effect if partial upgrading?
E.g., if some components call `AppStudioTokenProvider` while others call `M365TokenProvider`, will that causes multiple logins?

Reply:
Normally, all these logins use the same cache logic, they should share the login status.